src/dmd/expressionsem.d

```diff
@@ -5809,20 +5809,8 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             printf("AssertExp::semantic('%s')\n", exp.toChars());
         }
 
-        // save expression as a string before any semantic expansion
-        // if -checkaction=context is enabled an no message exists
         const generateMsg = !exp.msg && global.params.checkAction == CHECKACTION.context;
-        auto assertExpMsg = generateMsg ? exp.toChars() : null;
-
-        if (Expression ex = unaSemantic(exp, sc))
```

Comment:
This early expression sema of `exp.e1`, the assert condition, was one of the problems. E.g.:
```D
struct S
{
    int a;

    this(this) {}

    bool opEquals(S) @safe
    {
        return true;
    }
}

void foo()
{
    S a;
    assert(S() == a);
}
```

Here, an lvalue `a` is passed to `S.opEquals()`, which takes the arg by value. The early `CallExp` sema already lowers the argument expression `a` to a temporary, to something like `S __copytmp = (__copytmp = a).__postblit()`. *That* expression (instead of `a`) was previously stored as a further `__assertOp` temporary, and the original argument replaced by that temporary (without postblit).

A related 2nd issue concerns rvalue arguments - the early `CallExp` sema detects that no temporary is needed (move; elide postblit and dtor in the caller); the rvalue was previously correctly stored as `__assertOp` temporary, but as it is now an lvalue (possibly accessed a 2nd time for the `_d_assert_fail()` call if the assertion fails), it must be copied + postblitted. This can be fixed by deferring the `CallExp` sema until the arguments have been replaced by required temporaries.